src/utils/SpatialTreeInterface/SpatialTreeInterface.jl

```diff
@@ -0,0 +1,226 @@
+module SpatialTreeInterface
+
+import ..LoopStateMachine: @controlflow
+
+import Extents
+import GeoInterface as GI
+import AbstractTrees
+
+# ## Interface
+# Interface definition for spatial tree types.
+# There is no abstract supertype here since it's impossible to enforce,
+# but we do have a few methods that are common to all spatial tree types.
+
+"""
+    isspatialtree(tree)::Bool
+
+Return true if the object is a spatial tree, false otherwise.
+
+## Implementation notes
+
+For type stability, if your spatial tree type is `MyTree`, you should define
+`isspatialtree(::Type{MyTree}) = true`, and `isspatialtree(::MyTree)` will forward
+to that method automatically.
+"""
+isspatialtree(::T) where T = isspatialtree(T)
+isspatialtree(::Type{<: Any}) = false
+
+
+"""
+    getchild(node)
+
+Return an iterator over all the children of a node.
+This may be materialized if necessary or available,
+but can also be lazy (like a generator).
+"""
+getchild(node) = AbstractTrees.children(node)
+
+"""
+    getchild(node, i)
+
+Return the `i`-th child of a node.
+"""
+getchild(node, i) = getchild(node)[i]
+
+"""
+    nchild(node)
+
+Return the number of children of a node.
+"""
+nchild(node) = length(getchild(node))
+
+"""
+    isleaf(node)
+
+Return true if the node is a leaf node, i.e., there are no "children" below it.
+[`getchild`](@ref) should still work on leaf nodes, though, returning an iterator over the extents stored in the node - and similarly for `getnodes.`
+"""
+isleaf(node) = error("isleaf is not implemented for node type $(typeof(node))")
+
+"""
+    child_indices_extents(node)
+
+Return an iterator over the indices and extents of the children of a node.
+
+Each value of the iterator should take the form `(i, extent)`.
+"""
+function child_indices_extents(node)
+    return zip(1:nchild(node), getchild(node))
+end
+
+# ## Query functions
+# These are generic functions that work with any spatial tree type that implements the interface.
+
+
+"""
+    do_query(f, predicate, tree)
```

Comment:
Is there a reason both this and `query` are documented?

Isn't `do_query` internal?

Reply:
No, do_query is the non allocating version of query.   Maybe we can merge them though.

Do query also does not guarantee a sorted order of visits whereas query does.  For eg point in polygon we don't care about order but in poly intersection we do.

Reply:
No, do_query is the non allocating version of query.   Maybe we can merge them though.

Do query also does not guarantee a sorted order of visits whereas query does.  For eg point in polygon we don't care about order but in poly intersection we do.

Reply:
Maybe we can find a more specific verb than "do" to clarify the difference

Reply:
this is now `depth_first_search`